File: solutions/javascript/src/algorithm/queue/Exercise5.js
import DynamicQueueADT from "../../structure/dynamic/DynamicQueueADT.js";

function esCapicua(cola) {
  const valores = [];
  const auxiliar = new DynamicQueueADT();

  // Extraer elementos y guardarlos en una lista
  while (!cola.isEmpty()) {
    const valor = cola.getElement();
    valores.push(valor);
    auxiliar.add(valor);
    cola.remove();
  }

  // Restaurar la cola original
  while (!auxiliar.isEmpty()) {
    cola.add(auxiliar.getElement());
    auxiliar.remove();
  }

  // Verificar si los valores son capicúa
  const invertido = [...valores].reverse();
  return JSON.stringify(valores) === JSON.stringify(invertido);
}

// Ejemplo de uso
const cola = new DynamicQueueADT();
[1, 2, 3, 2, 1].forEach(valor => cola.add(valor));

console.log("¿Es capicúa?", esCapicua(cola)); // true

File: solutions/python/algorithm/queue/Exercise5.py
from structure.dynamic.DynamicQueueADT import DynamicQueueADT

def es_capicua(cola: DynamicQueueADT) -> bool:
    # Copiamos los valores de la cola en una lista
    valores = []
    auxiliar = DynamicQueueADT()

    while not cola.is_empty():
        valor = cola.getElement()
        valores.append(valor)
        auxiliar.add(valor)
        cola.remove()

    # Restauramos la cola original
    while not auxiliar.is_empty():
        cola.add(auxiliar.getElement())
        auxiliar.remove()

    # Verificamos si la lista es igual a su reversa
    return valores == valores[::-1]


# Ejemplo de uso:
cola = DynamicQueueADT()
for valor in [1, 2, 3, 2, 1]:
    cola.add(valor)

cola.mostrar()
print("¿Es capicúa?", es_capicua(cola))  # True
cola.mostrar()  # Se mantiene intacta
